src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -26,50 +25,34 @@ public sealed partial class JsonSerializerOptions
         private static JsonConverter[]? s_defaultFactoryConverters;
 
         // Stores the JsonTypeInfo factory, which requires unreferenced code and must be rooted by the reflection-based serializer.
-        private static Func<Type, JsonSerializerOptions, JsonTypeInfo>? s_typeInfoCreationFunc;
+        private static IJsonTypeInfoResolver? s_defaultTypeInfoResolver;
 
         [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
         [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
+        [MemberNotNull(nameof(s_defaultTypeInfoResolver))]
         private static void RootReflectionSerializerDependencies()
         {
             // s_typeInfoCreationFunc is the last field assigned.
             // Use it as the sentinel to ensure that all dependencies are initialized.
-            if (Volatile.Read(ref s_typeInfoCreationFunc) is null)
+            if (Volatile.Read(ref s_defaultTypeInfoResolver) is null)
             {
                 s_defaultSimpleConverters = GetDefaultSimpleConverters();
                 s_defaultFactoryConverters = GetDefaultFactoryConverters();
                 // Explicitly ensure that the previous fields are initialized along with this one.
-                Volatile.Write(ref s_typeInfoCreationFunc, CreateJsonTypeInfo);
+                // We also need to ensure that all threads use same instance of s_defaultTypeInfoResolver or
+                // otherwise EqualityComparer for CachingContext may fail even though
+                // two options might assign _typeInfoResolver to s_defaultTypeInfoResolver
+                Interlocked.CompareExchange(ref s_defaultTypeInfoResolver, new DefaultJsonTypeInfoResolver(mutable: false), null);
```

Comment:
Per feedback elsewhere, we can avoid using interlocked if we expose the default immutable resolver as a lazily instantiated singleton.

Reply:
same comment - unfortunately that makes it linker unfriendly although it technically would be better

Reply:
I don't think it makes any difference w.r.t. linkability. The main benefit is that you should be able to do something like
```C#
_typeInfoResolver = DefaultJsonTypeInfoResolver.Instance;
```
in the code paths that do use reflection without risk of multiple instances being used.